W-22800606-add-components-dm (#585)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-extension-plugin.adoc

image::dataweave-extension-logs.png[Mule DX DataWeave LS option to select in the drop-down menu of the Output]

== Scope Visibility

When you author DataWeave libraries (`.dwl` files that other DataWeave code imports), the DataWeave extension supports the visibility model introduced in DataWeave 2.12.0:

> * The keywords `private` and `internal` are highlighted as language keywords.

Reviewer: (This list isn't parallel and includes a lot of passive voice.)

Reviewer: (If you haven't yet, consider running a style check against the .md version of the SG.)

* The `@VisibleTo` annotation is recognized when applied to `internal` directives.
* The editor reports visibility violations inline and uses the same wording as the compiler (`... cannot be accessed from ... as it is private to the module ...`, `... is internal to the component ...`).

> * Auto-completion filters out directives that are not visible from the current file. For example, `private` helpers from another file will not appear in the suggestion list.

Author: ```suggestion
* Auto-completion filters out directives that aren't visible from the current file. For example, `private` helpers defined in another file don't appear in the suggestion list.
```

* `Go to definition` continues to work, and the editor warns when you navigate to a directive that isn't accessible from the current file.

These features apply to library projects only. If you write a Mule transformation mapping inline (a `%dw 2.0` expression in a Transform Message component, a Set Variable expression, or a DataWeave expression embedded in a connector), the `internal` keyword and `@VisibleTo` annotation have no practical effect. An inline mapping is a self-contained script that can't be imported from another inline mapping, so there is nothing to constrain. The keywords are still parsed, but you don't need them in mapping code.

The same rule applies to `.dwl` modules defined inside the Mule application. Every inline mapping and application-level module belongs to an unnamed component, so they can freely reference each other's `internal` directives. The `private` keyword still applies and restricts a directive to its declaring file regardless of component.

> For the full visibility model (what `private` and `internal` mean, what a component is, and how the component descriptor is generated by the DataWeave Maven plugin), see xref:dataweave::dataweave-scope-visibility.adoc[Scope Visibility in DataWeave].

Reviewer: ```suggestion
For the full visibility model, including definitions for `private`, `internal`, and components, and how the DataWeave Maven plugin generates the component descriptor, see xref:dataweave::dataweave-scope-visibility.adoc[].
```
